### PR TITLE
User book info

### PR DESCRIPTION
**Pull request by mrcuong84**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ def create
     if logging_user&.authenticate(params[:session][:password])
       log_in logging_user
       flash[:success] = t('.success_login')
+      # binding.pry
```
Comment (dinhvareal) [inline app/controllers/sessions_controller.rb]: Remove the line

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -2,9 +2,14 @@ class UsersController < ApplicationController
   before_action :validate_user, only: %i[show edit]
 
   def show
+    @books = current_user.books.includes(
+      :images,
+      :user
+    ).page(params[:page]).per(4)
```
Comment (dinhvareal) [inline app/controllers/users_controller.rb]: @books = current_user.books
                                     .order(created_at: :desc)
                                     .includes(:images, :user)
                                     .page(params[:page]).per(10)

Comment (dinhvareal) [reply app/controllers/users_controller.rb]: dot is align

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (mrcuong84) [reply app/controllers/sessions_controller.rb]: aready resolved

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (mrcuong84) [reply app/controllers/users_controller.rb]: fixed

File: config/locales/en.yml [hunk 3]
```diff
@@ -88,6 +92,7 @@ en:
       input_guide_password: "atleast 6 characters"
```
Comment (banguyenvareal) [inline config/locales/en.yml]: `at least`

File: config/locales/vi.yml [hunk 4]
```diff
@@ -73,7 +73,10 @@ vi:
       join: "Tham gia từ"
```
Comment (banguyenvareal) [inline config/locales/vi.yml]: should `join_from`

**Review by banguyenvareal — APPROVED** [review]

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -1,7 +1,93 @@
...
+      it 'flash and returns to home page' do
+        expect(flash[:danger]).to eql(I18n.t('not_found'))
+      end
+      it 'redirect to home' do
+        expect(subject).to redirect_to(root_url)
+      end
+    end
+    context 'current user' do
+      before do
+        100.times do
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: use a random integer

File: spec/controllers/users_controller_spec.rb [hunk 6]
```diff
@@ -1,7 +1,93 @@
...
+          @book = current_user.books.create!(
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: use factory bot

File: spec/controllers/users_controller_spec.rb [hunk 7]
```diff
@@ -1,7 +1,93 @@
...
+        }
+      end
+      it 'flash and returns to home page' do
+        expect(flash[:danger]).to eql(I18n.t('not_found'))
+      end
+      it 'redirect to home' do
+        expect(subject).to redirect_to(root_url)
+      end
+    end
+    context 'current user' do
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb:46]: missing test for user's info

File: spec/controllers/users_controller_spec.rb [hunk 8]
```diff
@@ -1,7 +1,93 @@
...
+            description: Faker::Lorem.sentence
+          )
+        end
+      end
+      it "quantity of book must equal quantity of user's book" do
+        expect(current_user.books.count).to eql(100)
+      end
+    end
+  end
+  describe 'create user' do
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: #create

File: spec/controllers/users_controller_spec.rb [hunk 9]
```diff
@@ -1,7 +1,93 @@
...
+    context 'with valid params' do
+      before do
+        post 'create', params: {
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: make sure that the params is correct

File: spec/controllers/users_controller_spec.rb [hunk 10]
```diff
@@ -1,7 +1,93 @@
...
+            password_confirmation: '123456',
+            avatar: ''
+          }
+        }
+      end
+      it 'flash success' do
+        expect(flash[:success]).to eql(I18n.t('.users.create.user_created'))
+      end
+      it 'redirect to created user' do
+        expect(subject).to redirect_to(user_path(id: user.id))
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: user.last or assigns(:user)

File: spec/controllers/users_controller_spec.rb [hunk 11]
```diff
@@ -1,7 +1,93 @@
...
+      end
+      it "attributes must comply with created user's attributes" do
+        expect(assigns(:user).attributes).to eql(user.attributes)
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: you must compare the user.last (user) and your input

File: spec/controllers/users_controller_spec.rb [hunk 12]
```diff
@@ -1,7 +1,93 @@
...
+      end
+
+      it 'the quantity of user increase 1' do
+        expect(User.count).to equal(user_count + 1)
+      end
+    end
+
+    context 'with invalid params' do
+      before do
+        post :create, params: {
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: same 40th line

**Review by banguyenvareal — APPROVED** [review]

File: spec/controllers/users_controller_spec.rb [hunk 13]
```diff
@@ -1,7 +1,107 @@
 require 'rails_helper'
-
 RSpec.describe UsersController, type: :controller do
   include_context 'logged in'
+  let(:other_user) { create(:user) }
+  let!(:user_count) { User.count }
+  let(:user) { User.last }
+  let(:n) { rand(100) }
```
Comment (hieunguyentrung) [inline spec/controllers/users_controller_spec.rb]: ???

File: spec/controllers/users_controller_spec.rb [hunk 14]
```diff
@@ -1,7 +1,107 @@
 require 'rails_helper'
-
 RSpec.describe UsersController, type: :controller do
   include_context 'logged in'
+  let(:other_user) { create(:user) }
+  let!(:user_count) { User.count }
+  let(:user) { User.last }
```
Comment (hieunguyentrung) [inline spec/controllers/users_controller_spec.rb]: Remove

File: spec/controllers/users_controller_spec.rb [hunk 15]
```diff
@@ -1,7 +1,107 @@
...
+      end
+      it 'redirect to created user' do
+        expect(subject).to redirect_to(user_path(id: User.last.id))
+      end
+      it "attributes must comply with created user's attributes" do
+        expect(assigns(:user).attributes).to eql(User.last.attributes)
+      end
+
+      it 'the quantity of user increase 1' do
+        expect(user_count + 1).to equal(User.count)
```
Comment (hieunguyentrung) [inline spec/controllers/users_controller_spec.rb]: Use expect change https://relishapp.com/rspec/rspec-expectations/v/2-0/docs/matchers/expect-change

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: spec/controllers/users_controller_spec.rb (hunk 14 above)
Comment (mrcuong84) [reply spec/controllers/users_controller_spec.rb]: resolved

File: spec/controllers/users_controller_spec.rb (hunk 13 above)
Comment (mrcuong84) [reply spec/controllers/users_controller_spec.rb]: resolved

File: app/controllers/sessions_controller.rb [hunk 16]
```diff
@@ -4,29 +4,29 @@ def new
...
+    redirect_to root_url
   end
 
   private
 
   def logging_user
     @logging_user ||= User.find_by(phone: params[:session][:phone])
   end
+
+  def login_sucess
```
Comment (hieunguyentrung) [inline app/controllers/sessions_controller.rb]: `valid_account?`

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,10 +1,25 @@
 class UsersController < ApplicationController
   before_action :validate_user, only: %i[show edit]
 
+  def index
+    if current_user
+      redirect_to current_user
+
```
Comment (hieunguyentrung) [inline app/controllers/users_controller.rb]: Remove.

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -1,10 +1,25 @@
 class UsersController < ApplicationController
   before_action :validate_user, only: %i[show edit]
 
+  def index
```
Comment (hieunguyentrung) [inline app/controllers/users_controller.rb]: Remove.

File: app/views/users/show.html.erb [hunk 19]
```diff
@@ -18,10 +18,18 @@
...
       <div class="row">
         <a href="<%= edit_user_url %> " class="btn btn-primary"><%= t('.update') %></a>
       </div>
   </div>
 </div>
+
+<% if current_user.books.any? %>
+
+  <h3 class="text-center"><%= t('.selling_book_list') %></h3>
+<div class="col-md-12 book-col">
```
Comment (hieunguyentrung) [inline app/views/users/show.html.erb:30]: Indent.

File: spec/controllers/books/delete_spec.rb [hunk 20]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+RSpec.describe BooksController, type: :controller do
+  describe "#delete user's book" do
+    include_context 'logged in'
+    context 'current user remove book in the list' do
+      subject { delete :destroy, params: { id: Book.last.id } }
+      before do
+        rand(3..10).times do
+          @book = create(:book, user_id: current_user.id)
```
Comment (hieunguyentrung) [inline spec/controllers/books/delete_spec.rb]: Why do you need it ?

File: spec/controllers/sessions_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+          session: {
+            phone: user.phone,
+            password: other_password
+          }
+        }
+      end
+
+      it 'flash login fail' do
+        subject
+        expect(flash[:warning])
```
Comment (hieunguyentrung) [inline spec/controllers/sessions_controller_spec.rb:43]: Need to test `flash.count == 1`

File: spec/controllers/sessions_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    end
+  end
+
+  describe '#destroy' do
+    include_context 'logged in'
+    before do
+      session[:user_id] = current_user.id
+    end
+    context 'user log_out' do
+      subject { delete :destroy }
```
Comment (hieunguyentrung) [inline spec/controllers/sessions_controller_spec.rb:60]: Add new line below.

File: spec/controllers/sessions_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+      it 'session of current user will be removed' do
+        subject
+        expect(session[:user_id]).to eql(nil)
+      end
```
Comment (hieunguyentrung) [inline spec/controllers/sessions_controller_spec.rb:65]: Add new line below.

File: spec/controllers/users_controller_spec.rb [hunk 24]
```diff
@@ -1,7 +1,104 @@
...
+  end
+
+  let(:invalid_user_params) { { name: nil } }
+
+  describe '#show' do
+    context 'user have not logged in' do
+      subject { get :show, params: { id: other_user.id } }
+      before do
+        allow(User).to receive(:find_by).with(anything).and_return(nil)
+        subject
```
Comment (hieunguyentrung) [inline spec/controllers/users_controller_spec.rb]: Remove.

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 17 above)
Comment (mrcuong84) [reply app/controllers/users_controller.rb]: resolved

File: app/controllers/users_controller.rb (hunk 18 above)
Comment (mrcuong84) [reply app/controllers/users_controller.rb]: resolved

File: app/views/users/show.html.erb (hunk 19 above)
Comment (mrcuong84) [reply app/views/users/show.html.erb:30]: resolved

File: spec/controllers/sessions_controller_spec.rb (hunk 21 above)
Comment (mrcuong84) [reply spec/controllers/sessions_controller_spec.rb:43]: resolved

File: spec/controllers/sessions_controller_spec.rb (hunk 23 above)
Comment (mrcuong84) [reply spec/controllers/sessions_controller_spec.rb:65]: resolved

File: spec/controllers/sessions_controller_spec.rb (hunk 22 above)
Comment (mrcuong84) [reply spec/controllers/sessions_controller_spec.rb:60]: resolved

File: spec/controllers/books/delete_spec.rb (hunk 20 above)
Comment (mrcuong84) [reply spec/controllers/books/delete_spec.rb]: resolved

File: app/controllers/books_controller.rb [hunk 25]
```diff
@@ -37,6 +37,12 @@ def update
     end
   end
 
+  def destroy
+    Book.find(params[:id]).destroy
```
Comment (dinhvareal) [inline app/controllers/books_controller.rb]: Before action destroy is find_and_assign_book, you got @book

File: spec/controllers/sessions_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,77 @@
+require 'rails_helper'
+RSpec.describe SessionsController, type: :controller do
+  let(:password) { 'Aa123456' }
+  let(:other_password) { 'Aa00000' }
+  let(:user) do
+    create(:user, password: password, password_confirmation: password)
+  end
+
+  describe '#create' do
+    subject do
```
Comment (dinhvareal) [inline spec/controllers/sessions_controller_spec.rb]: move it into context

File: spec/controllers/users_controller_spec.rb [hunk 27]
```diff
@@ -1,7 +1,102 @@
...
+    context 'user have not logged in' do
+      subject { get :show, params: { id: other_user.id } }
+      before do
+        allow(User).to receive(:find_by).with(anything).and_return(nil)
+      end
+      it 'redirect to home' do
+        expect(subject).to redirect_to(root_url)
+      end
+
+      it 'flash and returns to home page' do
```
Comment (dinhvareal) [inline spec/controllers/users_controller_spec.rb]: rename the example

File: app/controllers/books_controller.rb (hunk 25 above)
Comment (mrcuong84) [reply app/controllers/books_controller.rb]: resolved

File: spec/controllers/sessions_controller_spec.rb (hunk 26 above)
Comment (mrcuong84) [reply spec/controllers/sessions_controller_spec.rb]: resolved

File: spec/controllers/users_controller_spec.rb (hunk 27 above)
Comment (mrcuong84) [reply spec/controllers/users_controller_spec.rb]: resolved

File: app/controllers/sessions_controller.rb [hunk 28]
```diff
@@ -4,24 +4,24 @@ def new
...
-    log_out
+    session.delete(:user_id)
+    @current_user = nil
     redirect_to root_url
   end
 
-  def log_out
-    session.delete(:user_id)
-    @current_user = nil
+  def login_sucess
```
Comment (hieunguyentrung) [inline app/controllers/sessions_controller.rb]: `valid_account?`

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: app/controllers/sessions_controller.rb (hunk 28 above)
Comment (mrcuong84) [reply app/controllers/sessions_controller.rb]: already change the name of the def to authenticate_sucess

**Review by hieunguyentrung — APPROVED** [review]

**Pull request merged**